include/SDL3/SDL_intrin.h

```diff
@@ -93,26 +93,42 @@ _m_prefetch(void *__P)
 #endif
 #endif /* compiler version */
 
+#if defined(__clang__) && defined(__has_attribute)
+# if __has_attribute(target)
+# define SDL_HAS_TARGET_ATTRIBS
+# endif
+#elif defined(__GNUC__) && (__GNUC__ + (__GNUC_MINOR__ >= 9) > 4) /* gcc >= 4.9 */
+# define SDL_HAS_TARGET_ATTRIBS
+#endif
+
+#ifdef SDL_HAS_TARGET_ATTRIBS
+# define SDL_TARGETING(x) __attribute__((target(x)))
+#else
+# define SDL_TARGETING(x)
+#endif
+
 #if defined(__loongarch_sx) && !defined(SDL_DISABLE_LSX)
 #include <lsxintrin.h>
 #endif
 #if defined(__loongarch_asx) && !defined(SDL_DISABLE_LASX)
 #include <lasxintrin.h>
 #endif
-#if defined(__AVX__) && !defined(SDL_DISABLE_AVX)
+#if defined(__x86_64__) || defined(_M_X64) || defined(__i386__) || defined(_M_IX86)
+#if (defined(__AVX__) || defined(SDL_HAS_TARGET_ATTRIBS)) && !defined(SDL_DISABLE_AVX)
```

Comment:
If the compiler does not have `SDL_HAS_TARGET_ATTRIBS`, then the function will only use intrinsics from the `imintrin.h` header. The compiler will not optimize the function further.

Reply:
Are `SDL_DISABLE_???` _always_ defined for non-x86?

Reply:
[With CMake, yes](https://github.com/libsdl-org/SDL/blob/08d5c1386e8eb468f0c66b5511c2d334e6ae0b22/CMakeLists.txt#L1007-L1037).
But that does not matter, because this location is not reached because of the `#iff defined(__x86_64__) || defined(_M_X64) || defined(__i386__) || defined(_M_IX86)`.  That `#ifdef` must remain. I was talking about the inner `#ifdef`.

Reply:
> [With CMake, yes](https://github.com/libsdl-org/SDL/blob/08d5c1386e8eb468f0c66b5511c2d334e6ae0b22/CMakeLists.txt#L1007-L1037).

OK, I see. I hope other solutions such as Android.mk, Xcode, etc don't have problems

> But that does not matter, because this location is not reached because of the `#if defined(__x86_64__) || defined(_M_X64) || defined(__i386__) || defined(_M_IX86)`. That `#ifdef` must remain. I was talking about the inner `#ifdef`.

OK, I see. However:

> If the compiler does not have `SDL_HAS_TARGET_ATTRIBS`, then the function will only use intrinsics from the `imintrin.h` header. The compiler will not optimize the function further.

If the compiler doesn't have target attributes and the relevant SIMD macro, `__SSE2__` for e.g., isn't defined, then the compiler will error out from within `#include <emmintrin.h>`

Reply:
Oh, that's weird.
This source compiles (and runs) file with arguments `-m32 -march=i386` (and doesn't print anything to stdout).
```c
#include <emmintrin.h>
#include <stdio.h>
int main(int argc, char *argv[]) {
#if defined(__SSE2__)
 printf("__SSE2__ defined\n");
#endif
 return 0;
}
```
About `emmintrin.h` throwing. In that case, the sse2 tests in the CMake script should have failed and `SDL_DISABLE_SSE2` should be defined.

About the cmake script, the lines below should be removed for all intel simd extensions. We want the inclusion tests to have the same compiler arguments as the the SDL sources.
https://github.com/libsdl-org/SDL/blob/08d5c1386e8eb468f0c66b5511c2d334e6ae0b22/CMakeLists.txt#L717-L718
https://github.com/libsdl-org/SDL/blob/08d5c1386e8eb468f0c66b5511c2d334e6ae0b22/CMakeLists.txt#L734

Reply:
> Oh, that's weird. This source compiles (and runs) file with arguments `-m32 -march=i386` (and doesn't print anything to stdout).

Because your compiler _does have_ target attributes and the emmintrin.h provided by that version of the compiler applies target attributes to all the inlines it defines. An emmintrin.h provided by a gcc or clang without target attributes _will_ error out like:
```
In file included from 0.c:1:
/usr/lib/gcc/i686-redhat-linux/4.4.7/include/emmintrin.h:32:3: error: #error "SSE2 instruction set not enabled"
```

Reply:
Yes, it does `#pragma GCC target("sse2")`.
If that is the behavior of the older toolchains, then this looks good :)

Reply:
OK

Reply:
I still want to hear your take in my `SDL_TARGET_xxx` suggestion for cmake

Reply:
I think these should be configured by the user through `CMAKE_C_FLAGS`. Users and packagers then only need to modify that argument when e.g. bumping the x86 feature level from `-march=x86-64` to `-march=x86-64-v3`.

Reply:
My only concern here is: up until today, we used to build the whole tree with -msse3 unless disabled explicitly. With this patch we may be downgrading the optimization level of non-SIMD parts of the source tree.

Reply:
Indeed, looks like that is part of `march=x86-64-v2` where most distros currently do `-march=x86-64`.

Looking at [its documentation] https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#ssetechs=SSE3,SSSE3), it only matters when handling integer arrays.
Does that apply to SDL in other locations outside of audio and video?

What benchmarks are used to measure SDL performance?

Reply:
> Indeed, looks like that is part of `march=x86-64-v2` where most distros currently do `-march=x86-64`.
> 
> Looking at [its documentation] https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#ssetechs=SSE3,SSSE3), it only matters when handling integer arrays. Does that apply to SDL in other locations outside of audio and video?

Don't know

Reply:
> Indeed, looks like that is part of `march=x86-64-v2` where most distros currently do `-march=x86-64`.

The effect of such downgrade will also apply to @slouken's own binary releases, which include i686-mingw i.e. non-x64

Reply:
Let's revert the cmake script to the original behavior until we measured the performance:
Add the `-mxxx` options when `SDL_XXX` is true and disable `SDL_AVX` by default

Reply:
No, please don’t do that. The use of intrinsics should be separate from the build flags, and the build flags should not be tied to intrinsic enablement. Please leave off any architecture specific build flags and let the distribution packagers set those appropriately for their hardware targets.

Reply:
Okay then.
Can you give more info about performance measuring?
It might be useful to integrate into cmake.

Reply:
OK, can you document how to insert additional switches to CFLAGS somewhere
for cmake noobs? `CMAKE_C_FLAGS=xxx cmake [other stuff]` or something like it?